<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Emma Rand">

<title>Workshop: programmatic protein structure analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="antibody_mimetics_workshop_3_files/libs/clipboard/clipboard.min.js"></script>
<script src="antibody_mimetics_workshop_3_files/libs/quarto-html/quarto.js"></script>
<script src="antibody_mimetics_workshop_3_files/libs/quarto-html/popper.min.js"></script>
<script src="antibody_mimetics_workshop_3_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="antibody_mimetics_workshop_3_files/libs/quarto-html/anchor.min.js"></script>
<link href="antibody_mimetics_workshop_3_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="antibody_mimetics_workshop_3_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="antibody_mimetics_workshop_3_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="antibody_mimetics_workshop_3_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="antibody_mimetics_workshop_3_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Workshop: programmatic protein structure analysis</h1>
<p class="subtitle lead">BIO00088H - Group Research Project</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Emma Rand </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="based-on-the-antibody-mimetics-workshop-by-michael-plevin-jon-agirre" class="level1">
<h1>Based on the <a href="https://colab.research.google.com/drive/1BOmOi9e_yHeFXucBnYwUCJmFfZLBRFKf">Antibody Mimetics Workshop</a> by Michael Plevin &amp; Jon Agirre</h1>
<section id="understanding-this-document-in-rstudio" class="level2">
<h2 class="anchored" data-anchor-id="understanding-this-document-in-rstudio">Understanding this document in RStudio</h2>
<p>You are using RStudio which is an integrated development environment (IDE) initially developed for R. It can now be used for many other languages including Python. Some set is is required which is covered in <a href="https://3mmarand.github.io/BIO00088H-data/structure/structures.html">Structure Data Analysis for Group Project</a>. This in contrast to the Colab notebook which is cloud-based and does not require any set up (except installing packages)</p>
<p>You might find RStudio useful for Python because you are already familiar with it. It is also a very good way to create <a href="https://3mmarand.github.io/BIO00088H-data/core/week-11/overview.html">Quarto documents</a> with code chunks in more than one language. This document is an example of that.</p>
<p>For example, this is an R code chunk. You can run it by clicking on the green arrow in the top right of the chunk. Or do <code>Cmd+Return</code> on a Mac or <code>Ctrl+Return</code> on Windows.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "C:/Users/er13/OneDrive - University of York/Desktop/Desktop/undergrad-teaching-york/BIO00088H/structure-analysis"</code></pre>
</div>
</div>
<p>Notice:</p>
<ul>
<li>prompt in the console is <code>&gt;</code> as you are used to in R.</li>
<li>The Environment Pane is showing an R environment</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/r-chunk-in-quarto.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Screenshot of RStudio highlighting that the prompt is an R prompt (&gt;) and the environment is the R environment</figcaption><p></p>
</figure>
</div>
<p>This is a Python code chunk. You can also run it by clicking on the green arrow in the top right of the chunk. Or do <code>Cmd+Return</code> on a Mac or <code>Ctrl+Return</code> on Windows.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>os.getcwd()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'C:\\Users\\er13\\OneDrive - University of York\\Desktop\\Desktop\\undergrad-teaching-york\\BIO00088H\\structure-analysis'</code></pre>
</div>
</div>
<p>Notice:</p>
<ul>
<li>the R command <code>reticulate::repl_python()</code> was sent to the console to start Python</li>
<li>prompt in the console is <code>&gt;&gt;&gt;</code>. This is the Python prompt. If you try typing R code like <code>getwd()</code> in the console at the <code>&gt;&gt;&gt;</code> prompt you will get an error.</li>
<li>The Environment Pane is showing a Python environment containing an ‘interface’ to the r environment. This is the <code>r</code> object.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/py-chunk-in-quarto.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Screenshot of RStudio highlighting that the prompt is a Python prompt (&gt;&gt;&gt;) and the environment is the Python environment containing an object r</figcaption><p></p>
</figure>
</div>
<p>If you type <code>r.getwd()</code> in the console at the Python prompt you don’t get an error. You get the same result as you did in the R code chunk above</p>
<p>This is because you can access R commands and R objects from your Python environment using the <code>r</code> object. For example, you can access an R vector object <code>an_r_obj</code> from Python using <code>r.an_r_obj</code>.</p>
<p>Run this code to create a vector in R called <code>an_r_obj</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>an_r_obj <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(an_r_obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> num [1:3] 1 2 3</code></pre>
</div>
</div>
<p>You’ll see <code>quit</code> is sent to the Python prompt to exit Python and run the R command.</p>
<p>Try to access it from Python:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>an_r_obj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>name 'an_r_obj' is not defined</code></pre>
</div>
</div>
<p>You get an error because Python does not know about the R object <code>an_r_obj</code>. But you <em>can</em> access it using the <code>r</code> object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>r.an_r_obj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1.0, 2.0, 3.0]</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="bu">type</span>(r.an_r_obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'list'&gt;</code></pre>
</div>
</div>
<p>You can ‘copy’ it to your python environment if you want:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>acopy <span class="op">=</span> r.an_r_obj</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="bu">type</span>(acopy)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;class 'list'&gt;</code></pre>
</div>
</div>
<p><code>acopy</code> is in your Python environment now.</p>
<p>Magic!</p>
<p>Datatypes differ between the languages. Reticulate converts some automatically - for example R vectors to Python lists. For some other datatypes you must explicitly convert.</p>
<p>You can also access Python objects from R using the <code>py</code> object. For example, you can access the Python list object <code>a_python_obj</code> from R using <code>py$a_python_obj</code>. However you do need to have loaded the <code>reticulate</code> package in R first.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>a_python_obj <span class="op">=</span> [<span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>py<span class="sc">$</span>a_python_obj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4 5 6</code></pre>
</div>
</div>
<ul>
<li><p>You access objects in the Python environment from R in an R-like way using <code>py$</code></p></li>
<li><p>You access objects in the R environment from Python in a Python-like way using <code>r.</code></p></li>
</ul>
</section>
<section id="getting-started" class="level2">
<h2 class="anchored" data-anchor-id="getting-started">Getting started</h2>
<p>This RStudio Project already has a folder containing the <code>4INS.pdb</code> file. I have called the folder <code>pdb_files</code>. Jon called his folder <code>download</code>. When you are doing your own work in RStudio you can make folders by:</p>
<ul>
<li>clicking on the New Folder button in the Files Pane OR</li>
<li>typing <code>mkdir pdb_files</code> in the <em>Terminal</em> OR</li>
<li>typing <code>fs::dir_create("pdb_files")</code> at the R console</li>
</ul>
<p>You can download a pdb by typing <code>curl https://files.rcsb.org/download/4INS.pdb -o pdb_files/4INS.pdb</code> in the <em>Terminal</em></p>
<p>I have used <code>curl</code> rather than <code>wget</code> because <code>wget</code> because most macOS and Windows do not come with <code>wget</code> by default. They mostly do have the very similar <code>curl</code>. Use is very similar but notice it is a lowercase <code>-o</code> for output rather than a capital <code>-O</code>.</p>
<p>I would tend not to have this “house-keeping” (making folders, downloading through the terminal) in my reproducible document. I would, however, explain the directory structure and where the data files came from in the README.md file. Later we download data using a Python command. I probably would have that in my reproducible document.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
</section>
<section id="analysis-code" class="level2">
<h2 class="anchored" data-anchor-id="analysis-code">Analysis code</h2>
<p>Most of the text explanation here is taken directly from Jon and Michael’s document.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import the `gemmi` package</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> gemmi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>4INS is a 1.5 angstroems structure of insulin. We have seen structures of insulin in our first workshop.</p>
<p>You can open the file to view it by clicking on it in the Files Pane.</p>
<p>Here are a few basic but interesting questions about a protein structure we can answer using Python programming:</p>
<ul>
<li>What is the molecular weight of the protein?</li>
<li>What is the modelled sequence? We know the UniPro sequence, but is it the same as the one in the crystal?</li>
<li>What is the average B-factor per-residue? Would be neat to separate by main-chain and side-chain atoms.</li>
<li>Does the protein have any disulphide bonds?</li>
</ul>
<p>Import the structure file:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> <span class="st">"pdb_files/4INS.pdb"</span> </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>structure_file <span class="op">=</span> gemmi.read_structure ( filename ) <span class="co"># remember gemmi.whatever?</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(structure_file) <span class="co"># to delete</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;gemmi.Structure 4INS with 1 model(s)&gt;</code></pre>
</div>
</div>
<p>Extract the first model from the structure object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> structure_file[<span class="dv">0</span>] <span class="co"># there can be more than one model per structure file!</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>                          <span class="co"># in computer science 0 is the first index, so we are</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>                          <span class="co"># getting the first (and only) model in the file</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(model) <span class="co"># to delete</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;gemmi.Model 1 with 4 chain(s)&gt;</code></pre>
</div>
</div>
<p>You can see all the attributes and methods available for the model object with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="bu">dir</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>['__class__', '__delattr__', '__delitem__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'add_chain', 'all', 'calculate_center_of_mass', 'calculate_mass', 'clone', 'count_atom_sites', 'count_hydrogen_sites', 'count_occupancies', 'find_chain', 'find_cra', 'find_last_chain', 'find_residue_group', 'get_all_residue_names', 'get_subchain', 'has_hydrogen', 'name', 'remove_alternative_conformations', 'remove_chain', 'remove_hydrogens', 'remove_ligands_and_waters', 'remove_waters', 'sole_residue', 'split_chains_by_segments', 'subchains', 'transform_pos_and_adp']</code></pre>
</div>
</div>
<p>You can also click on it in the Environment Pane.</p>
<p>One of the methods calculates the molecular weight of the protein:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> ( <span class="st">"This protein weight is"</span>, model.calculate_mass(), <span class="st">"Da"</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>This protein weight is 15332.740152545452 Da</code></pre>
</div>
</div>
<p>Which can be formatted to two decimal places:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> ( <span class="st">"This protein weight is </span><span class="sc">%.2f</span><span class="st"> Da"</span> <span class="op">%</span> model.calculate_mass() )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>This protein weight is 15332.74 Da</code></pre>
</div>
</div>
<p>Each chain in the model can be accessed by indexing the model object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(model[<span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;gemmi.Chain A with 52 res&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(model[<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;gemmi.Chain B with 80 res&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(model[<span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;gemmi.Chain C with 28 res&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(model[<span class="dv">3</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;gemmi.Chain D with 294 res&gt;</code></pre>
</div>
</div>
<p>You can iterate through the chains in the model with a for loop:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> polymer <span class="kw">in</span> model :</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span> (polymer)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;gemmi.Chain A with 52 res&gt;
&lt;gemmi.Chain B with 80 res&gt;
&lt;gemmi.Chain C with 28 res&gt;
&lt;gemmi.Chain D with 294 res&gt;</code></pre>
</div>
</div>
<p>You can iterate through the chains and through the residues in the chain with two nested for loops:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> polymer <span class="kw">in</span> model :</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span> (polymer)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> steve <span class="kw">in</span> polymer :</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(steve)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;gemmi.Chain A with 52 res&gt;
1(GLY)
2(ILE)
3(VAL)
4(GLU)
5(GLN)
6(CYS)
7(CYS)
8(THR)
9(SER)
10(ILE)
11(CYS)
12(SER)
13(LEU)
14(TYR)
15(GLN)
16(LEU)
17(GLU)
18(ASN)
19(TYR)
20(CYS)
21(ASN)
101(HOH)
102(HOH)
103(HOH)
104(HOH)
105(HOH)
106(HOH)
107(HOH)
108(HOH)
109(HOH)
110(HOH)
111(HOH)
112(HOH)
113(HOH)
114(HOH)
115(HOH)
116(HOH)
117(HOH)
118(HOH)
119(HOH)
120(HOH)
121(HOH)
122(HOH)
123(HOH)
124(HOH)
125(HOH)
126(HOH)
127(HOH)
128(HOH)
129(HOH)
130(HOH)
131(HOH)
&lt;gemmi.Chain B with 80 res&gt;
1(PHE)
2(VAL)
3(ASN)
4(GLN)
5(HIS)
6(LEU)
7(CYS)
8(GLY)
9(SER)
10(HIS)
11(LEU)
12(VAL)
13(GLU)
14(ALA)
15(LEU)
16(TYR)
17(LEU)
18(VAL)
19(CYS)
20(GLY)
21(GLU)
22(ARG)
23(GLY)
24(PHE)
25(PHE)
26(TYR)
27(THR)
28(PRO)
29(LYS)
30(ALA)
101(ZN)
201(HOH)
202(HOH)
203(HOH)
204(HOH)
205(HOH)
206(HOH)
207(HOH)
208(HOH)
209(HOH)
210(HOH)
211(HOH)
212(HOH)
213(HOH)
214(HOH)
215(HOH)
216(HOH)
217(HOH)
218(HOH)
219(HOH)
220(HOH)
221(HOH)
222(HOH)
223(HOH)
224(HOH)
225(HOH)
226(HOH)
227(HOH)
228(HOH)
229(HOH)
230(HOH)
231(HOH)
232(HOH)
233(HOH)
234(HOH)
235(HOH)
236(HOH)
237(HOH)
238(HOH)
239(HOH)
240(HOH)
241(HOH)
242(HOH)
243(HOH)
244(HOH)
245(HOH)
246(HOH)
247(HOH)
248(HOH)
249(HOH)
&lt;gemmi.Chain C with 28 res&gt;
1(GLY)
2(ILE)
3(VAL)
4(GLU)
5(GLN)
6(CYS)
7(CYS)
8(THR)
9(SER)
10(ILE)
11(CYS)
12(SER)
13(LEU)
14(TYR)
15(GLN)
16(LEU)
17(GLU)
18(ASN)
19(TYR)
20(CYS)
21(ASN)
101(HOH)
102(HOH)
103(HOH)
104(HOH)
105(HOH)
106(HOH)
107(HOH)
&lt;gemmi.Chain D with 294 res&gt;
1(PHE)
2(VAL)
3(ASN)
4(GLN)
5(HIS)
6(LEU)
7(CYS)
8(GLY)
9(SER)
10(HIS)
11(LEU)
12(VAL)
13(GLU)
14(ALA)
15(LEU)
16(TYR)
17(LEU)
18(VAL)
19(CYS)
20(GLY)
21(GLU)
22(ARG)
23(GLY)
24(PHE)
25(PHE)
26(TYR)
27(THR)
28(PRO)
29(LYS)
30(ALA)
101(ZN)
201(HOH)
202(HOH)
203(HOH)
204(HOH)
205(HOH)
206(HOH)
207(HOH)
208(HOH)
209(HOH)
210(HOH)
211(HOH)
212(HOH)
213(HOH)
214(HOH)
215(HOH)
216(HOH)
217(HOH)
218(HOH)
219(HOH)
220(HOH)
221(HOH)
222(HOH)
223(HOH)
224(HOH)
225(HOH)
226(HOH)
227(HOH)
228(HOH)
229(HOH)
230(HOH)
231(HOH)
232(HOH)
233(HOH)
234(HOH)
235(HOH)
236(HOH)
237(HOH)
238(HOH)
239(HOH)
240(HOH)
241(HOH)
242(HOH)
243(HOH)
244(HOH)
245(HOH)
246(HOH)
247(HOH)
248(HOH)
249(HOH)
250(HOH)
251(HOH)
252(HOH)
253(HOH)
254(HOH)
255(HOH)
256(HOH)
257(HOH)
258(HOH)
259(HOH)
260(HOH)
261(HOH)
262(HOH)
263(HOH)
264(HOH)
265(HOH)
266(HOH)
267(HOH)
268(HOH)
269(HOH)
270(HOH)
271(HOH)
272(HOH)
273(HOH)
274(HOH)
275(HOH)
276(HOH)
277(HOH)
278(HOH)
279(HOH)
280(HOH)
281(HOH)
282(HOH)
283(HOH)
284(HOH)
285(HOH)
286(HOH)
287(HOH)
288(HOH)
289(HOH)
290(HOH)
291(HOH)
292(HOH)
293(HOH)
294(HOH)
295(HOH)
296(HOH)
297(HOH)
298(HOH)
299(HOH)
300(HOH)
301(HOH)
302(HOH)
303(HOH)
304(HOH)
305(HOH)
306(HOH)
307(HOH)
308(HOH)
309(HOH)
310(HOH)
311(HOH)
312(HOH)
313(HOH)
314(HOH)
315(HOH)
316(HOH)
317(HOH)
318(HOH)
319(HOH)
320(HOH)
321(HOH)
322(HOH)
323(HOH)
324(HOH)
325(HOH)
326(HOH)
327(HOH)
328(HOH)
329(HOH)
330(HOH)
331(HOH)
332(HOH)
333(HOH)
334(HOH)
335(HOH)
336(HOH)
337(HOH)
338(HOH)
339(HOH)
340(HOH)
341(HOH)
342(HOH)
343(HOH)
344(HOH)
345(HOH)
346(HOH)
347(HOH)
348(HOH)
349(HOH)
350(HOH)
351(HOH)
352(HOH)
353(HOH)
354(HOH)
355(HOH)
356(HOH)
357(HOH)
358(HOH)
359(HOH)
360(HOH)
361(HOH)
362(HOH)
363(HOH)
364(HOH)
365(HOH)
366(HOH)
367(HOH)
368(HOH)
369(HOH)
370(HOH)
371(HOH)
372(HOH)
373(HOH)
374(HOH)
375(HOH)
376(HOH)
377(HOH)
378(HOH)
379(HOH)
380(HOH)
381(HOH)
382(HOH)
383(HOH)
384(HOH)
385(HOH)
386(HOH)
387(HOH)
388(HOH)
389(HOH)
390(HOH)
391(HOH)
392(HOH)
393(HOH)
394(HOH)
395(HOH)
396(HOH)
397(HOH)
398(HOH)
399(HOH)
400(HOH)
401(HOH)
402(HOH)
403(HOH)
404(HOH)
405(HOH)
406(HOH)
407(HOH)
408(HOH)
409(HOH)
410(HOH)
411(HOH)
412(HOH)
413(HOH)
414(HOH)
415(HOH)
416(HOH)
417(HOH)
418(HOH)
419(HOH)
420(HOH)
421(HOH)
422(HOH)
423(HOH)
424(HOH)
425(HOH)
426(HOH)
427(HOH)
428(HOH)
429(HOH)
430(HOH)
431(HOH)
432(HOH)
433(HOH)
434(HOH)
435(HOH)
436(HOH)
437(HOH)
438(HOH)
439(HOH)
440(HOH)
441(HOH)
442(HOH)
443(HOH)
444(HOH)
445(HOH)
446(HOH)
447(HOH)
448(HOH)
449(HOH)
450(HOH)
451(HOH)
452(HOH)
453(HOH)
454(HOH)
455(HOH)
456(HOH)
457(HOH)
458(HOH)
459(HOH)
460(HOH)
461(HOH)
462(HOH)
463(HOH)</code></pre>
</div>
</div>
<p>A major difference between R and Python is that R uses various braces to indicate blocks of code. The indentation makes the code easier to follow. In Python the indentation is not just for readability, it is part of the syntax. It is the indentation itself which indicates the blocks of code.</p>
<p>You can use to nested loops to iterate through the chains and residues and extra and print the chain name and the one letter code for the amino acid:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> polymer <span class="kw">in</span> model :</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  sequence <span class="op">=</span> <span class="st">""</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span> (polymer)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> steve <span class="kw">in</span> polymer :</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>    sequence <span class="op">+=</span> gemmi.find_tabulated_residue(steve.name).one_letter_code.upper()</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span> (<span class="st">"Polymer </span><span class="sc">%s</span><span class="st">: </span><span class="sc">%s</span><span class="st">"</span> <span class="op">%</span> (polymer.name,sequence))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;gemmi.Chain A with 52 res&gt;
Polymer A: GIVEQCCTSICSLYQLENYCN                               
&lt;gemmi.Chain B with 80 res&gt;
Polymer B: FVNQHLCGSHLVEALYLVCGERGFFYTPKA                                                  
&lt;gemmi.Chain C with 28 res&gt;
Polymer C: GIVEQCCTSICSLYQLENYCN       
&lt;gemmi.Chain D with 294 res&gt;
Polymer D: FVNQHLCGSHLVEALYLVCGERGFFYTPKA                                                                                                                                                                                                                                                                        </code></pre>
</div>
</div>
<p>The <code>+=</code> operator is a short-hand way of writing <code>sequence = sequence + gemmi.find_tabulated_residue(steve.name).one_letter_code.upper()</code>. It is a common operator in many programming languages. It grows the string <code>sequence</code> by adding the one letter code for the amino acid each time the inner loops executes.</p>
<p>We are going to use our new skills for calculating average B-factors for this structure. We will make a crude separation between protein and waters, and we will introduce another loop to go through every atom in every residue. We will rely on the fact that the <code>find_tabulated_residue()</code> function does not return a code for monomers/residues that are not amino acids. In that case, we will be using <code>residue.name</code> to check if the monomer we are checking is a water molecule.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statistics        <span class="co"># we will need to do stats, so we import the module</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>aminoacid_bfactors <span class="op">=</span> [ ] <span class="co"># these are lists where we are going to store the</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>rest_bfactors      <span class="op">=</span> [ ] <span class="co"># individual B-factors</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chain <span class="kw">in</span> model :</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> monomer <span class="kw">in</span> chain : <span class="co"># we'll call it monomer this time (name doesn't matter)</span></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> atom <span class="kw">in</span> monomer :</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> gemmi.find_tabulated_residue(monomer.name).is_amino_acid() :</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>        aminoacid_bfactors.append ( atom.b_iso )</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>      <span class="cf">else</span> :</span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>        rest_bfactors.append ( atom.b_iso )</span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> ( <span class="st">"Protein part mean(stdev) : </span><span class="sc">%.2f</span><span class="st"> (</span><span class="sc">%.2f</span><span class="st">)"</span> <span class="op">%</span> <span class="op">\</span></span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>      ( statistics.mean ( aminoacid_bfactors ), statistics.stdev ( aminoacid_bfactors ) ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Protein part mean(stdev) : 21.45 (12.39)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span> ( <span class="st">"Rest of atoms mean(stdev): </span><span class="sc">%.2f</span><span class="st"> (</span><span class="sc">%.2f</span><span class="st">)"</span> <span class="op">%</span> <span class="op">\</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>      ( statistics.mean ( rest_bfactors ), statistics.stdev ( rest_bfactors ) ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rest of atoms mean(stdev): 52.02 (16.73)</code></pre>
</div>
</div>
<p>Hopefully the code is more or less self-explanatory. The key bit in this new part is that we have used the <code>is_amino_acid()</code> function to check what type of monomer we’re encountering at each step. You might wonder, how do we come up with all those suggestions for wonderful functions that do exactly what we need? Well, have a look at the <a href="https://project-gemmi.github.io/python-api/gemmi.ResidueInfo.html">docs</a> for the <code>gemmi</code> module.</p>
<p>At this point you might be starting to want to wonder how to relate the notion of ‘Class’ to the names we have been using here so far. We will try to explain this as easy as possible.</p>
<p>We define a variable just by naming it, for example <code>aminoacid_bfactors</code> is a list because we have assigned <code>[ ]</code> – an empty list – to it in <code>aminoacid_bfactors = [ ]</code>.</p>
<p>Other variables are assigned whatever functions return, such as s<code>tructure_file = gemmi.read_structure ( filename )</code> earlier in the notebook. <code>gemmi.read_structure()</code> returns a Structure class, as it says on the documentation for that function.</p>
<p>Once you’ve got a variable of the <code>Structure</code> class, you can start invoking the functions that come with it.</p>
<p>🔨 Have a deep dive into the docs, see what functions are available and think about potential uses for them. Discuss those ideas with your neighbours and tell us about them! 😃</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we do not need to import gemmi or statistics again because the notebook</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remembers the previous block where we have imported the module and</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="co"># read the structure file</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chain <span class="kw">in</span> model :</span>
<span id="cb48-9"><a href="#cb48-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-10"><a href="#cb48-10" aria-hidden="true" tabindex="-1"></a>  main_chain_bfactors <span class="op">=</span> [ ] <span class="co"># we will store per residue averages here</span></span>
<span id="cb48-11"><a href="#cb48-11" aria-hidden="true" tabindex="-1"></a>  side_chain_bfactors <span class="op">=</span> [ ]</span>
<span id="cb48-12"><a href="#cb48-12" aria-hidden="true" tabindex="-1"></a>  residue_numbers     <span class="op">=</span> [ ] <span class="co"># residue 1 could be reside 42 according to UniProt</span></span>
<span id="cb48-13"><a href="#cb48-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-14"><a href="#cb48-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> monomer <span class="kw">in</span> chain : <span class="co"># we'll call it monomer this time (name doesn't matter)</span></span>
<span id="cb48-15"><a href="#cb48-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-16"><a href="#cb48-16" aria-hidden="true" tabindex="-1"></a>    monomer_main_bfactor <span class="op">=</span> [ ] <span class="co"># one for main-chains, one for side-chains</span></span>
<span id="cb48-17"><a href="#cb48-17" aria-hidden="true" tabindex="-1"></a>    monomer_side_bfactor <span class="op">=</span> [ ]</span>
<span id="cb48-18"><a href="#cb48-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> gemmi.find_tabulated_residue(monomer.name).is_amino_acid() :</span>
<span id="cb48-19"><a href="#cb48-19" aria-hidden="true" tabindex="-1"></a>      <span class="co"># we only want per-residue information on amino acids</span></span>
<span id="cb48-20"><a href="#cb48-20" aria-hidden="true" tabindex="-1"></a>      residue_numbers.append ( <span class="bu">int</span>(monomer.seqid.num) )</span>
<span id="cb48-21"><a href="#cb48-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-22"><a href="#cb48-22" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> atom <span class="kw">in</span> monomer : <span class="co"># EDIT HERE! You can find the answers with Moorhen.</span></span>
<span id="cb48-23"><a href="#cb48-23" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> ( atom.name <span class="op">==</span> <span class="st">"CA"</span> <span class="kw">or</span> atom.name <span class="op">==</span> <span class="st">"C"</span> <span class="kw">or</span></span>
<span id="cb48-24"><a href="#cb48-24" aria-hidden="true" tabindex="-1"></a>              atom.name <span class="op">==</span> <span class="st">"N"</span> <span class="kw">or</span> atom.name <span class="op">==</span> <span class="st">"O"</span> ) :</span>
<span id="cb48-25"><a href="#cb48-25" aria-hidden="true" tabindex="-1"></a>          monomer_main_bfactor.append ( atom.b_iso ) <span class="co"># clue: main chain atoms</span></span>
<span id="cb48-26"><a href="#cb48-26" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span> :</span>
<span id="cb48-27"><a href="#cb48-27" aria-hidden="true" tabindex="-1"></a>          monomer_side_bfactor.append ( atom.b_iso )</span>
<span id="cb48-28"><a href="#cb48-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-29"><a href="#cb48-29" aria-hidden="true" tabindex="-1"></a>      main_chain_bfactors.append ( statistics.mean ( monomer_main_bfactor ) )</span>
<span id="cb48-30"><a href="#cb48-30" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> <span class="bu">len</span> ( monomer_side_bfactor ) <span class="op">&gt;</span> <span class="dv">0</span> :</span>
<span id="cb48-31"><a href="#cb48-31" aria-hidden="true" tabindex="-1"></a>        side_chain_bfactors.append ( statistics.mean ( monomer_side_bfactor ) )</span>
<span id="cb48-32"><a href="#cb48-32" aria-hidden="true" tabindex="-1"></a>      <span class="cf">else</span> :</span>
<span id="cb48-33"><a href="#cb48-33" aria-hidden="true" tabindex="-1"></a>        side_chain_bfactors.append ( <span class="fl">0.0</span> ) <span class="co"># special case for glycine</span></span>
<span id="cb48-34"><a href="#cb48-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-35"><a href="#cb48-35" aria-hidden="true" tabindex="-1"></a>  plt.title (<span class="st">"</span><span class="ch">\n</span><span class="st">B-factor plot for chain "</span> <span class="op">+</span> chain.name,</span>
<span id="cb48-36"><a href="#cb48-36" aria-hidden="true" tabindex="-1"></a>             fontsize <span class="op">=</span> <span class="dv">15</span>,</span>
<span id="cb48-37"><a href="#cb48-37" aria-hidden="true" tabindex="-1"></a>             fontweight <span class="op">=</span> <span class="st">"bold"</span> )</span>
<span id="cb48-38"><a href="#cb48-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-39"><a href="#cb48-39" aria-hidden="true" tabindex="-1"></a>  plt.plot ( residue_numbers, main_chain_bfactors, <span class="st">"r--"</span>, label<span class="op">=</span><span class="st">"main chain"</span>)</span>
<span id="cb48-40"><a href="#cb48-40" aria-hidden="true" tabindex="-1"></a>  plt.bar ( residue_numbers, side_chain_bfactors, alpha<span class="op">=</span><span class="fl">0.4</span>, label<span class="op">=</span><span class="st">"side chain"</span> )</span>
<span id="cb48-41"><a href="#cb48-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-42"><a href="#cb48-42" aria-hidden="true" tabindex="-1"></a>  plt.xlabel ( <span class="st">"Residue"</span> )</span>
<span id="cb48-43"><a href="#cb48-43" aria-hidden="true" tabindex="-1"></a>  plt.ylabel ( <span class="st">"Isotropic B-factor"</span> )</span>
<span id="cb48-44"><a href="#cb48-44" aria-hidden="true" tabindex="-1"></a>  plt.ylim ( <span class="fl">0.0</span> )</span>
<span id="cb48-45"><a href="#cb48-45" aria-hidden="true" tabindex="-1"></a>  plt.legend ( )</span>
<span id="cb48-46"><a href="#cb48-46" aria-hidden="true" tabindex="-1"></a>  plt.show ( )</span>
<span id="cb48-47"><a href="#cb48-47" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span> (<span class="st">"</span><span class="ch">\n\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="antibody_mimetics_workshop_3_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="antibody_mimetics_workshop_3_files/figure-html/unnamed-chunk-24-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="antibody_mimetics_workshop_3_files/figure-html/unnamed-chunk-24-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="antibody_mimetics_workshop_3_files/figure-html/unnamed-chunk-24-4.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The code above looks complicated, but it’s just an extension of what we had done previously, plus a bit of code at the end for plotting the B-factors. This last bit of code uses matplotlib, which is great for producing graphs in Python. You could tweak these plots to no end!</p>
<p>🔨 Customise the look of your B-factor plots using this<a href="https://colab.research.google.com/github/amanchadha/aman-ai/blob/master/matplotlib.ipynb#scrollTo=il1nR5kXSmQ9">really good Colab notebook</a> on the use of matplotlib as a reference. If you get stuck, ask one of us for help 🙂</p>
<p>💡 One potential improvement of these plots would be to draw a horizontal line where the Wilson B is. You can find out its value on the PDB validation report, and you will learn how to draw horizontal lines in the next example anyway.</p>
<p>And on the subject of customisation, it would be great if we could do the same for AlphaFold models. We know that AlphaFold stores its pLDDT in the B-factor column, so perhaps it’s doable with very little effort?</p>
<p>🔨 Run the following code to get a nice plot for an AlphaFold prediction for human Lysozyme (UniProt ID A0A080YUZ9). This time we will be getting an AlphaFold model from the database, and we will use Python’s wget module for that. We used the wget command at the start of the workshop (scroll up and have a look) and now we are going to demonstrate how we can use it directly from Python. But before that, we need to install it using dear old pip</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we do not need to import gemmi or statistics again because the notebook</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remembers the previous block where we have imported the module and</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="co"># read the structure file</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt <span class="co"># need to create a new plot</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> wget</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> wget.download(<span class="st">"https://alphafold.ebi.ac.uk/files/AF-A0A080YUZ9-F1-model_v4.pdb"</span>, <span class="st">"pdb_files"</span>)</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>structure_file <span class="op">=</span> gemmi.read_structure ( filename )</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> structure_file[<span class="dv">0</span>]</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chain <span class="kw">in</span> model : <span class="co"># just one chain with non-multimer AlphaFold</span></span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a>  plddts          <span class="op">=</span> [ ] <span class="co"># we will store per residue pLDDT here</span></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a>  residue_numbers <span class="op">=</span> [ ]</span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> monomer <span class="kw">in</span> chain :</span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a>    residue_numbers.append ( <span class="bu">int</span>(monomer.seqid.num) )</span>
<span id="cb49-20"><a href="#cb49-20" aria-hidden="true" tabindex="-1"></a>    plddts.append (monomer[<span class="dv">0</span>].b_iso) <span class="co"># we just need to get the first atom's pLDDT</span></span>
<span id="cb49-21"><a href="#cb49-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-22"><a href="#cb49-22" aria-hidden="true" tabindex="-1"></a>plt.title (<span class="st">"</span><span class="ch">\n</span><span class="st">pLDDT plot for chain "</span> <span class="op">+</span> chain.name,</span>
<span id="cb49-23"><a href="#cb49-23" aria-hidden="true" tabindex="-1"></a>            fontsize <span class="op">=</span> <span class="dv">15</span>,</span>
<span id="cb49-24"><a href="#cb49-24" aria-hidden="true" tabindex="-1"></a>            fontweight <span class="op">=</span> <span class="st">"bold"</span> )</span>
<span id="cb49-25"><a href="#cb49-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-26"><a href="#cb49-26" aria-hidden="true" tabindex="-1"></a>plt.plot ( residue_numbers, plddts, <span class="st">"r--"</span>, label<span class="op">=</span><span class="st">"pLDDT"</span>)</span>
<span id="cb49-27"><a href="#cb49-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-28"><a href="#cb49-28" aria-hidden="true" tabindex="-1"></a>plt.xlabel ( <span class="st">"Residue"</span> )</span>
<span id="cb49-29"><a href="#cb49-29" aria-hidden="true" tabindex="-1"></a>plt.ylabel ( <span class="st">"pLDDT"</span> )</span>
<span id="cb49-30"><a href="#cb49-30" aria-hidden="true" tabindex="-1"></a>plt.ylim ( <span class="fl">0.0</span>, <span class="fl">100.0</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(0.0, 100.0)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>plt.axhline(y<span class="op">=</span><span class="fl">90.0</span>, color<span class="op">=</span><span class="st">"blue"</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>plt.axhline(y<span class="op">=</span><span class="fl">70.0</span>, color<span class="op">=</span><span class="st">"turquoise"</span>)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>plt.axhline(y<span class="op">=</span><span class="fl">50.0</span>, color<span class="op">=</span><span class="st">"gold"</span>)</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>plt.axhline(y<span class="op">=</span><span class="fl">40.0</span>, color<span class="op">=</span><span class="st">"orange"</span>)</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>plt.legend ( )</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>plt.show ( )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="antibody_mimetics_workshop_3_files/figure-html/unnamed-chunk-25-9.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The code is actually simpler than what we had for the main and side chain B-factor averages. This time we only care about per residue numbers. We have added a number of horizontal lines coloured according to the <a href="https://www.alphafold.ebi.ac.uk/entry/A0A080YUZ9">AlphaFold database entry</a> so you can see where pLDDT gets too dodgy.</p>
<p>🔨 Go to the <a href="https://www.alphafold.ebi.ac.uk/entry/A0A080YUZ9">AlphaFold database entry</a> page and compare your results with the available graphics, including the Predicted Aligned Error (PAE). What is going on with the N-terminus? Discuss.</p>
<p>💡 Remember what we said about AlphaFold not doing post-translational modifications? That includes the ever-present disulphide bonds. The structure we are looking at is a Lysozyme, which as you will remember from past workshops, relies on disulphide bonds to form the final fold. Wouldn’t it be great if we could check if, despite not being formally linked in the PDB/mmCIF file, the cysteines in this AlphaFold model are consistent with the presence of disulphide bonds?</p>
<p>Here is some new code that checks for that.</p>
<p>🔨 You’ll have to complete it though!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we do not need to import gemmi or statistics again because the notebook</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remembers the previous block where we have imported the module and</span></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a><span class="co"># read the structure file</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>distance <span class="op">=</span> <span class="fl">3.0</span> <span class="co"># change for a sensible value!</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chain <span class="kw">in</span> model : <span class="co"># just one chain with non-multimer AlphaFold</span></span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>  neighbour_search <span class="op">=</span> gemmi.NeighborSearch ( model,</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>                                            structure_file.cell,</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>                                            distance ).populate ( include_h<span class="op">=</span><span class="va">False</span> )</span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a>  cysteines <span class="op">=</span> gemmi.Selection ( <span class="st">'(CYS)'</span> ) <span class="co"># fill in the three-letter code for cysteine!</span></span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> model <span class="kw">in</span> cysteines.models(structure_file) :</span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> chain <span class="kw">in</span> cysteines.chains(model) :</span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> residue <span class="kw">in</span> cysteines.residues(chain) :</span>
<span id="cb52-16"><a href="#cb52-16" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span> (<span class="st">"Checking "</span> <span class="op">+</span> residue.name <span class="op">+</span> <span class="st">" </span><span class="sc">%i</span><span class="st">"</span> <span class="op">%</span> residue.seqid.num )</span>
<span id="cb52-17"><a href="#cb52-17" aria-hidden="true" tabindex="-1"></a>        marks <span class="op">=</span> neighbour_search.find_neighbors ( residue.sole_atom(<span class="st">"SG"</span>), <span class="co"># name of sulphur atom in cysteine</span></span>
<span id="cb52-18"><a href="#cb52-18" aria-hidden="true" tabindex="-1"></a>                                                  <span class="dv">0</span>, distance )</span>
<span id="cb52-19"><a href="#cb52-19" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> mark <span class="kw">in</span> marks :</span>
<span id="cb52-20"><a href="#cb52-20" aria-hidden="true" tabindex="-1"></a>          cra <span class="op">=</span> mark.to_cra ( model )</span>
<span id="cb52-21"><a href="#cb52-21" aria-hidden="true" tabindex="-1"></a>          <span class="cf">if</span> cra.residue.seqid.num <span class="op">!=</span> residue.seqid.num :</span>
<span id="cb52-22"><a href="#cb52-22" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span> ( <span class="st">"Potential disulphide bond found: "</span> <span class="op">+</span> cra.residue.name,</span>
<span id="cb52-23"><a href="#cb52-23" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" </span><span class="sc">%i</span><span class="st">"</span> <span class="op">%</span> cra.residue.seqid.num, <span class="st">" to "</span> <span class="op">+</span> residue.name <span class="op">+</span></span>
<span id="cb52-24"><a href="#cb52-24" aria-hidden="true" tabindex="-1"></a>                    <span class="st">" </span><span class="sc">%i</span><span class="st">"</span> <span class="op">%</span> residue.seqid.num )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Checking CYS 8
Checking CYS 40
Checking CYS 58
Checking CYS 71
Potential disulphide bond found: CYS  191  to CYS 71
Checking CYS 95
Potential disulphide bond found: CYS  179  to CYS 95
Checking CYS 129
Potential disulphide bond found: CYS  144  to CYS 129
Checking CYS 140
Potential disulphide bond found: CYS  158  to CYS 140
Checking CYS 144
Potential disulphide bond found: CYS  129  to CYS 144
Checking CYS 158
Potential disulphide bond found: CYS  140  to CYS 158
Checking CYS 179
Potential disulphide bond found: CYS  95  to CYS 179
Checking CYS 191
Potential disulphide bond found: CYS  71  to CYS 191</code></pre>
</div>
</div>
<p>In this bit of code we are finding cysteines in a protein’s sequence with <code>gemmi.Selection()</code>, then searching for neighbours within a cut-off distance (your pick) from the position of their sulphur atoms (which you’ll need to identify using Moorhen or other tools), and then reporting the ones that don’t include atoms from the original cysteine, as the search is atom-based.</p>
<p>🔨 Check that your code does what it’s supposed to be doing. Use Moorhen to download A0A080YUZ9 and go to the positions reported by your code, check that the cysteines are within sensible bonding distance and in the expected orientation.</p>
<p>💡 You probably have spotted that the code reports every potential pair of cysteines twice. That’s one thing you could try to address if you were looking to improve this code!</p>
</section>
<section id="bonus-calculation-of-torsion-angles" class="level2">
<h2 class="anchored" data-anchor-id="bonus-calculation-of-torsion-angles">Bonus: calculation of torsion angles</h2>
<p>Here is one more bit of code for you to play with if you have finished all the previous steps and still want to learn more. We will calculate the three torsion angles involved in the formation of the protein backbone for a high resolution crystallographic structure – a monomer of thaumatin at 0.9 angstroems – plotting omega (should be close to planar) and phi,psi (ramachandran angles) separately.</p>
<p>🔨 Lots to customise on these plots: from the appearance of the data to what numbers are represented, all up to you to improve.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we do not need to import gemmi or statistics again because the notebook</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="co"># remembers the previous block where we have imported the module and</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="co"># read the structure file</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt <span class="co"># need to create a new plot</span></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> math <span class="im">import</span> degrees, isnan</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> wget</span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a>filename <span class="op">=</span> wget.download(<span class="st">"https://files.rcsb.org/download/5X9L.pdb"</span>, <span class="st">"pdb_files"</span>)</span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-12"><a href="#cb54-12" aria-hidden="true" tabindex="-1"></a>structure_file <span class="op">=</span> gemmi.read_structure ( filename )</span>
<span id="cb54-13"><a href="#cb54-13" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> structure_file[<span class="dv">0</span>]</span>
<span id="cb54-14"><a href="#cb54-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-15"><a href="#cb54-15" aria-hidden="true" tabindex="-1"></a>residue_numbers <span class="op">=</span> [ ]</span>
<span id="cb54-16"><a href="#cb54-16" aria-hidden="true" tabindex="-1"></a>omega_angles <span class="op">=</span> [ ]</span>
<span id="cb54-17"><a href="#cb54-17" aria-hidden="true" tabindex="-1"></a>phi_angles <span class="op">=</span> [ ]</span>
<span id="cb54-18"><a href="#cb54-18" aria-hidden="true" tabindex="-1"></a>psi_angles <span class="op">=</span> [ ]</span>
<span id="cb54-19"><a href="#cb54-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-20"><a href="#cb54-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chain <span class="kw">in</span> model :</span>
<span id="cb54-21"><a href="#cb54-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> residue <span class="kw">in</span> chain :</span>
<span id="cb54-22"><a href="#cb54-22" aria-hidden="true" tabindex="-1"></a>    next_res <span class="op">=</span> chain.next_residue ( residue )</span>
<span id="cb54-23"><a href="#cb54-23" aria-hidden="true" tabindex="-1"></a>    prev_res <span class="op">=</span> chain.previous_residue ( residue )</span>
<span id="cb54-24"><a href="#cb54-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> next_res:</span>
<span id="cb54-25"><a href="#cb54-25" aria-hidden="true" tabindex="-1"></a>      omega <span class="op">=</span> gemmi.calculate_omega(residue, next_res)</span>
<span id="cb54-26"><a href="#cb54-26" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> <span class="kw">not</span> isnan(degrees(omega)):</span>
<span id="cb54-27"><a href="#cb54-27" aria-hidden="true" tabindex="-1"></a>        omega_angles.append ( degrees ( omega ) )</span>
<span id="cb54-28"><a href="#cb54-28" aria-hidden="true" tabindex="-1"></a>      residue_numbers.append ( residue.seqid.num )</span>
<span id="cb54-29"><a href="#cb54-29" aria-hidden="true" tabindex="-1"></a>    phi, psi <span class="op">=</span> gemmi.calculate_phi_psi ( prev_res, residue, next_res )</span>
<span id="cb54-30"><a href="#cb54-30" aria-hidden="true" tabindex="-1"></a>    phi_angles.append ( phi )</span>
<span id="cb54-31"><a href="#cb54-31" aria-hidden="true" tabindex="-1"></a>    psi_angles.append ( psi )</span>
<span id="cb54-32"><a href="#cb54-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-33"><a href="#cb54-33" aria-hidden="true" tabindex="-1"></a>plt.title (<span class="st">"</span><span class="ch">\n</span><span class="st">Omega plot"</span>,</span>
<span id="cb54-34"><a href="#cb54-34" aria-hidden="true" tabindex="-1"></a>             fontsize <span class="op">=</span> <span class="dv">15</span>,</span>
<span id="cb54-35"><a href="#cb54-35" aria-hidden="true" tabindex="-1"></a>             fontweight <span class="op">=</span> <span class="st">"bold"</span> )</span>
<span id="cb54-36"><a href="#cb54-36" aria-hidden="true" tabindex="-1"></a><span class="co"># counts, bins = np.histogram ( omega_angles )</span></span>
<span id="cb54-37"><a href="#cb54-37" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.stairs ( counts, bins )</span></span>
<span id="cb54-38"><a href="#cb54-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-39"><a href="#cb54-39" aria-hidden="true" tabindex="-1"></a>plt.hist(omega_angles, bins<span class="op">=</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>(array([ 75.,   4.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   1.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   1.,  12., 113.]), array([-1.79948164e+02, -1.67957104e+02, -1.55966045e+02, -1.43974985e+02,
       -1.31983925e+02, -1.19992865e+02, -1.08001805e+02, -9.60107455e+01,
       -8.40196857e+01, -7.20286259e+01, -6.00375661e+01, -4.80465063e+01,
       -3.60554465e+01, -2.40643867e+01, -1.20733269e+01, -8.22670542e-02,
        1.19087928e+01,  2.38998526e+01,  3.58909124e+01,  4.78819722e+01,
        5.98730320e+01,  7.18640918e+01,  8.38551516e+01,  9.58462114e+01,
        1.07837271e+02,  1.19828331e+02,  1.31819391e+02,  1.43810451e+02,
        1.55801510e+02,  1.67792570e+02,  1.79783630e+02]), &lt;BarContainer object of 30 artists&gt;)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Angle / Degrees"</span>)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"Frequency"</span>)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>plt.show ( )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="antibody_mimetics_workshop_3_files/figure-html/unnamed-chunk-27-11.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"</span><span class="ch">\n\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>plt.title (<span class="st">"</span><span class="ch">\n</span><span class="st">Ramachandran angles"</span>,</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>             fontsize <span class="op">=</span> <span class="dv">15</span>,</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>             fontweight <span class="op">=</span> <span class="st">"bold"</span> )</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>plt.scatter ( phi_angles, psi_angles )</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>plt.show ( )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="antibody_mimetics_workshop_3_files/figure-html/unnamed-chunk-27-12.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> session_info</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>session_info.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>-----
gemmi               0.6.3
matplotlib          3.5.1
numpy               1.21.5
session_info        1.0.0
wget                3.2
-----
Python 3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]
Windows-10-10.0.19045-SP0
-----
Session information updated at 2023-11-23 12:18</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">session_info</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.3.1 (2023-06-16 ucrt)
 os       Windows 10 x64 (build 19045)
 system   x86_64, mingw32
 ui       RTerm
 language (EN)
 collate  English_United Kingdom.utf8
 ctype    English_United Kingdom.utf8
 tz       Europe/London
 date     2023-11-23
 pandoc   3.1.1 @ C:/Program Files/RStudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 package     * version date (UTC) lib source
 cachem        1.0.7   2023-02-24 [1] CRAN (R 4.3.1)
 callr         3.7.3   2022-11-02 [1] CRAN (R 4.3.1)
 cli           3.6.1   2023-03-23 [1] CRAN (R 4.3.1)
 crayon        1.5.2   2022-09-29 [1] CRAN (R 4.3.1)
 devtools      2.4.5   2022-10-11 [1] CRAN (R 4.3.1)
 digest        0.6.31  2022-12-11 [1] CRAN (R 4.3.1)
 ellipsis      0.3.2   2021-04-29 [1] CRAN (R 4.3.1)
 evaluate      0.21    2023-05-05 [1] CRAN (R 4.3.1)
 fastmap       1.1.1   2023-02-24 [1] CRAN (R 4.3.1)
 fs            1.6.1   2023-02-06 [1] CRAN (R 4.3.1)
 glue          1.6.2   2022-02-24 [1] CRAN (R 4.3.1)
 here          1.0.1   2020-12-13 [1] CRAN (R 4.3.1)
 htmltools     0.5.5   2023-03-23 [1] CRAN (R 4.3.1)
 htmlwidgets   1.6.2   2023-03-17 [1] CRAN (R 4.3.1)
 httpuv        1.6.9   2023-02-14 [1] CRAN (R 4.3.1)
 jsonlite      1.8.7   2023-06-29 [1] CRAN (R 4.3.1)
 knitr         1.43    2023-05-25 [1] CRAN (R 4.3.1)
 later         1.3.0   2021-08-18 [1] CRAN (R 4.3.1)
 lattice       0.20-45 2021-09-22 [1] CRAN (R 4.3.1)
 lifecycle     1.0.3   2022-10-07 [1] CRAN (R 4.3.1)
 magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.3.1)
 Matrix        1.5-3   2022-11-11 [1] CRAN (R 4.3.1)
 memoise       2.0.1   2021-11-26 [1] CRAN (R 4.3.1)
 mime          0.12    2021-09-28 [1] CRAN (R 4.3.0)
 miniUI        0.1.1.1 2018-05-18 [1] CRAN (R 4.3.1)
 pkgbuild      1.4.0   2022-11-27 [1] CRAN (R 4.3.1)
 pkgload       1.3.2   2022-11-16 [1] CRAN (R 4.3.1)
 png           0.1-8   2022-11-29 [1] CRAN (R 4.3.0)
 prettyunits   1.1.1   2020-01-24 [1] CRAN (R 4.3.1)
 processx      3.8.0   2022-10-26 [1] CRAN (R 4.3.1)
 profvis       0.3.7   2020-11-02 [1] CRAN (R 4.3.1)
 promises      1.2.0.1 2021-02-11 [1] CRAN (R 4.3.1)
 ps            1.7.3   2023-03-21 [1] CRAN (R 4.3.1)
 purrr         1.0.1   2023-01-10 [1] CRAN (R 4.3.1)
 R6            2.5.1   2021-08-19 [1] CRAN (R 4.3.1)
 Rcpp          1.0.10  2023-01-22 [1] CRAN (R 4.3.1)
 remotes       2.4.2   2021-11-30 [1] CRAN (R 4.3.1)
 reticulate  * 1.30    2023-06-09 [1] CRAN (R 4.3.1)
 rlang         1.1.0   2023-03-14 [1] CRAN (R 4.3.1)
 rmarkdown     2.23    2023-07-01 [1] CRAN (R 4.3.1)
 rprojroot     2.0.3   2022-04-02 [1] CRAN (R 4.3.1)
 rstudioapi    0.14    2022-08-22 [1] CRAN (R 4.3.1)
 sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.3.1)
 shiny         1.7.4.1 2023-07-06 [1] CRAN (R 4.3.1)
 stringi       1.7.12  2023-01-11 [1] CRAN (R 4.3.0)
 stringr       1.5.0   2022-12-02 [1] CRAN (R 4.3.1)
 urlchecker    1.0.1   2021-11-30 [1] CRAN (R 4.3.1)
 usethis       2.1.6   2022-05-25 [1] CRAN (R 4.3.1)
 vctrs         0.6.1   2023-03-22 [1] CRAN (R 4.3.1)
 withr         2.5.0   2022-03-03 [1] CRAN (R 4.3.1)
 xfun          0.39    2023-04-20 [1] CRAN (R 4.3.1)
 xtable        1.8-4   2019-04-21 [1] CRAN (R 4.3.1)
 yaml          2.3.7   2023-01-23 [1] CRAN (R 4.3.0)

 [1] C:/Users/er13/AppData/Local/R/win-library/4.3
 [2] C:/Program Files/R/R-4.3.1/library

─ Python configuration ───────────────────────────────────────────────────────
 python:         C:/ProgramData/Anaconda3/python.exe
 libpython:      C:/ProgramData/Anaconda3/python39.dll
 pythonhome:     C:/ProgramData/Anaconda3
 version:        3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]
 Architecture:   64bit
 numpy:          C:/ProgramData/Anaconda3/Lib/site-packages/numpy
 numpy_version:  1.21.5
 
 NOTE: Python version was forced by RETICULATE_PYTHON_FALLBACK

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
</section>
</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>The decision on whether to download data from the internet before an analysis, or download each time the analysis runs, depends on the circumstances. If the data definitely will not be updated it does not really matter. If they do get updated you would want to download before and record the date if you are writing a publication but download fresh each time if you are providing a live update like web application.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>